File: examples/typescript/src/webhook-selftest.ts
// Runs the shared vectors (examples/testdata/webhook-vectors.json) through verifyWebhook.
import { readFileSync } from "node:fs";
import { verifyWebhook } from "./webhook.ts";

interface Vector { name: string; secret: string; id: string; timestamp: number; body: string; signature: string; now: number; valid: boolean }
const { cases } = JSON.parse(readFileSync(new URL("../../testdata/webhook-vectors.json", import.meta.url), "utf8")) as { cases: Vector[] };

let failed = 0;
for (const c of cases) {
  let valid = true;
  try {
    verifyWebhook(c.secret, { "webhook-id": c.id, "webhook-timestamp": String(c.timestamp), "webhook-signature": c.signature }, c.body, c.now);
  } catch {
    valid = false;
  }
  if (valid !== c.valid) {
    failed++;
    console.error(`FAIL ${c.name}: got valid=${valid}, want ${c.valid}`);
  }
}
console.log(`typescript webhook: ${cases.length - failed}/${cases.length} vectors`);
process.exit(failed ? 1 : 0);

File: examples/csharp/WebhookSelfTest.cs
// Runs the shared vectors (examples/testdata/webhook-vectors.json) through DonaWebhook.Verify.
using System.Text;
using System.Text.Json;

namespace DonaExamples;

public static class WebhookSelfTest
{
    public static int Run()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "testdata", "webhook-vectors.json");
        using var doc = JsonDocument.Parse(File.ReadAllText(path));
        int total = 0, failed = 0;
        foreach (var c in doc.RootElement.GetProperty("cases").EnumerateArray())
        {
            total++;
            bool valid = true;
            try
            {
                DonaWebhook.Verify(c.GetProperty("secret").GetString()!, c.GetProperty("id").GetString()!,
                    c.GetProperty("timestamp").GetInt64().ToString(), c.GetProperty("signature").GetString()!,
                    Encoding.UTF8.GetBytes(c.GetProperty("body").GetString()!),
                    DateTimeOffset.FromUnixTimeSeconds(c.GetProperty("now").GetInt64()));
            }
            catch (DonaWebhookVerificationException) { valid = false; }
            if (valid != c.GetProperty("valid").GetBoolean())
            {
                failed++;
                Console.Error.WriteLine($"FAIL {c.GetProperty("name").GetString()}: got valid={valid}");
            }
        }
        Console.WriteLine($"csharp webhook: {total - failed}/{total} vectors");
        return failed == 0 ? 0 : 1;
    }
}
